akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -242,27 +242,47 @@ private[stream] object Collect {
   override def initialAttributes: Attributes = DefaultAttributes.collect and SourceLocation.forLambda(pf)
 
   def createLogic(inheritedAttributes: Attributes) =
-    new SupervisedGraphStageLogic(inheritedAttributes, shape) with InHandler with OutHandler {
-
-      import Collect.NotApplied
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+      private var buffer: OptionVal[Out] = OptionVal.none
 
-      val wrappedPf = () => pf.applyOrElse(grab(in), NotApplied)
+      override def preStart(): Unit = pull(in)
 
-      override def onPush(): Unit = withSupervision(wrappedPf) match {
-        case Some(result) =>
-          result match {
-            case NotApplied             => pull(in)
-            case result: Out @unchecked => push(out, result)
-            case _                      => throw new RuntimeException() // won't happen, compiler exhaustiveness check pleaser
+      def onPush(): Unit =
+        try {
+          val elem = grab(in)
+          import Collect.NotApplied
+          pf.applyOrElse(elem, NotApplied) match {
+            case NotApplied => pull(in)
+            case result: Out @unchecked =>
+              if (isAvailable(out)) {
+                push(out, result)
+                pull(in)
+              } else {
+                buffer = OptionVal.Some(result)
+              }
+            case _ => throw new RuntimeException() // won't happen, compiler exhaustiveness check pleaser
           }
-        case None => //do nothing
-      }
+        } catch {
+          case NonFatal(ex) =>
+            decider(ex) match {
+              case Supervision.Stop => failStage(ex)
+              case _                => if (!isClosed(in)) pull(in) else completeStage()
```

Comment:
previously it also checked hasBeenPulled. not needed now?

Reply:
I think so, the reviouse `hasBeenPulled` is never tested in the test too.

Reply:
yes, @johanandren  so should the filter be updated too,  the filter have a big try catch too.

Reply:
For filter operator:
```scala
      override def onPush(): Unit = {
        val elem = grab(in)
        try if (p(elem)) {
          if (isAvailable(out)) {
            push(out, elem)
            pull(in)
          } else {
            buffer = OptionVal.Some(elem)
            contextPropagation.suspendContext()
          }
        } else {
          pull(in)
        } catch {
          case NonFatal(ex) =>
            decider(ex) match {
              case Supervision.Stop => failStage(ex)
              case _                => pull(in)
            }
        }
      }
```

Reply:
On second thought, maybe it is fine as is, I realized how much trickier the logic will be with a try-catch first and then try to handle the value outside of the try catch, probably not worth it.

Reply:
Let me update it when arrive home.